design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,172 @@
+
+
+## Synchronization Approach
+
+Synchronization must handle two scenarios: (i) a label is added/removed on a BMH and must be added/removed on the corresponding Node and (ii) a label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on BMH. 
+
+The actual implementation of the synchronization logic may be encapsulated into an additional controller running within CAPM3; however, in the discussion below we simply reference CAPM3 as the entity doing the work.
+
+Scenario (i) can be handled as follows:
+
+
+
+1. CAPM3 will set up a watch on the BMH resource. It only requires read-only access similar to what is done today:
+
+    baremetalhosts,verbs=get;list;watch
+
+2. On each BMH reconcile event, the controller will fetch the corresponding Node. The _mapping_ can be accomplished by:
+
+		BMH.ConsumerRef --> Metal3Machine.OwnerRef --> Machine.NodeRef
+
+
+
+3. Synchronize the labels and update the Node in the workload cluster. CAPM3 already fetches the Node objects from the workload clusters today using a remote client. We would leverage that same capability here. 
+
+An alternative means to accomplish the mapping is to use the ProviderID: 
+
+    BMH.ConsumerRef --> Metal3Machine.Spec.ProviderID --> Node.Spec.ProviderID
+
+Unless we can set up filters, the approach would require us to fetch all Nodes and then scan through them to find the Node matching the ProviderID. This does not seem scalable for large clusters. [Field selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/) could potentially be employed and would require further investigation. However, there is no obvious advantage over the earlier approach for doing this mapping. 
+
+Scenario (ii) can be handled with a full-sync of all BMH and Node objects at a preconfigured synchronization interval:
+
+
+
+1. At a configurable interval (e.g. 15 seconds), fetch the complete list of BMHs.
+2. For each BMH, fetch the corresponding Node using the mapping discussed above. 
+3. Synchronize the labels and update the Node in the workload cluster.
+
+An alternative would be to set up a watch on the Node resource types in every workload cluster. The benefit being that we can quickly synchronize any label changes on the Node object itself. However, this introduces complications when the management cluster != workload cluster and additionally, when there are many workload clusters being managed by a single management cluster. How such a watch could be implemented within the controller-runtime model would require further investigation. 
+
+
+## Configuration
+
+We need to expose two additional configuration variables to the user: the set of label prefixes and the full-sync interval. We can introduce two new command-line flags for CAPM3, and also utilize default values if the flags are not specified:
```

Comment:
User-configurable values like this should really be managed through an API, not command line options. Maybe in this case we could use an annotation on the host?

Reply:
My concern would be consistency since the same annotation would have to be duplicated across BMHs. The prefix, while user configurable, is a static value that should not require re-configuring. I would imagine that a cluster admin would define a standard set of prefixes to be used internally.

Also, annotations can be (accidentally) added/removed/changed. We could address this by having an immutable field in the BMH just for specifying prefixes.

Reply:
OpenShift cluster administrators do not deploy metal3 themselves and do not have access to the command line arguments for the cluster-api provider. So, command line arguments won't work at all in an OpenShift cluster.

I suggested an annotation on the host because it wasn't obvious that we would always want the same set of labels copied. If we do think it will always be the same set, then maybe a new API resource to describe which labels to copy would make sense. That would let the implementation for copying them be based on a new controller within CAPM3 that could reconcile on the new API type. If we had an API that offered a list of label prefixes and then a labelSelector to pick which hosts it applies to, that would let users have a way to copy all or some of the labels from all or some of the hosts.

Reply:
I have updated the proposal. 
Also, see the section on using Metal3Cluster. Could this be a good place to keep this info?